File: python/leetcode/304.py
class NumMatrix(object):
    def __init__(self, matrix):
        if not matrix or not matrix[0]: return
        m, n = len(matrix), len(matrix[0])
        self.dp = [[0 for j in xrange(n + 1)] for i in xrange(m + 1)]
        for i in xrange(1, m + 1):
            for j in xrange(1, n + 1):
                self.dp[i][j] = matrix[i - 1][j - 1] + self.dp[i][j - 1] + \
                self.dp[i - 1][j] - self.dp[i - 1][j - 1]

    def sumRegion(self, row1, col1, row2, col2):
        row1, col1, row2, col2 = row1 + 1, col1 + 1, row2 + 1, col2 + 1
        return self.dp[row2][col2] - self.dp[row2][col1 - 1] - \
        self.dp[row1 - 1][col2] + self.dp[row1 - 1][col1 - 1]

File: cpp/leetcode/304.cpp
class NumMatrix {
public:
    NumMatrix(vector<vector<int>> &matrix) {
        int m = matrix.size();
        int n = (m > 0) ? matrix[0].size() : 0;
        dp = vector<vector<int>>(m + 1, vector<int>(n + 1, 0));
        for(int i = 1; i <= m; ++i){
            for(int j = 1; j <= n; ++j){
                dp[i][j] = matrix[i-1][j-1] + dp[i][j-1] + dp[i-1][j] - dp[i-1][j-1];
            }
        }
        
    }

    int sumRegion(int row1, int col1, int row2, int col2) {
        return dp[row2+1][col2+1] - dp[row2+1][col1] - dp[row1][col2+1] + dp[row1][col1];
    }
private:
    vector<vector<int>> dp;
};